Subclass RCTEventEmitter in RCTUIManager

joshyhargreaves opened the pull request:
Updating RCTUIManager to use new events system and subclass RCTEventEmitter and bring it into line with other native modules; this fixes the deprecation warnings from calling [_bridge sendDeviceEventWithName...]. 

Fixes warnings in https://github.com/facebook/react-native/issues/11736

I am presuming that we would rather take this approach than explicitly use -Wdeprecated-declarations?

**Test plan (required)**
I'm not sure how to test this other than running tests and manual testing?
Can we write a test for this?

rh389 commented:
Thanks for looking at this. Just a couple more things to address I think -

1. The [mechanism for listening for module-scoped events is different](http://facebook.github.io/react-native/releases/0.42/docs/native-modules-ios.html#sending-events-to-javascript), so technically this is a BC - title should probably include that.
2. Does this trigger runtime warnings on dimension changes (eg screen rotation?). I think it's recommended that you only send events if you know there are listeners (see [here](http://facebook.github.io/react-native/releases/0.42/docs/native-modules-ios.html#optimizing-for-zero-listeners)).
3. [`Dimensions`](https://github.com/facebook/react-native/blob/master/Libraries/Utilities/Dimensions.js#L87) will need to be fixed. This might be a start:

```js
Dimensions.set(UIManager.Dimensions);

if (Platform.OS === 'ios') {
  const uiManagerEmitter = new NativeEventEmitter(UIManager);
  uiManagerEmitter.addListener('didUpdateDimensions', Dimensions.set);
} else {
  DeviceEventEmitter.addListener('didUpdateDimensions', Dimensions.set);
}
```

joshyhargreaves commented:
Thanks for providing some context on the module scoped events, and Dimensions, I'm AFK at the moment but will update later.

Regarding the runtime warnings on screen-rotation, I'll confirm, but I did briefly see that RCTEventEmitter had this logic within to only send the events if there were listeners in `sendEventWithName`

Some wider comments, regarding what you mentioned in the other issue about support for module scoped events on iOS and lack of support on Android. Do we want to diverge the APIs in this manner? Or like you said, do we just go ahead and wait for Android to catch up?

rh389 commented:
I think we're both right re events without listeners - there is logic to not send, but it will log a warning: https://github.com/facebook/react-native/blob/master/React/Modules/RCTEventEmitter.m#L54.

Re Android/iOS divergence - that's a question for the RN team, but I'd be inclined to do as you've done. It seems like there are obvious benefits to scoping events, especially when it comes to third-party modules, but I haven't seen any indication that there's any plan to bring this to android. We could be waiting forever and IMO we might as well improve what we can for now.

janicduplessis commented:
I don't think you need to do any platform forking with NativeEventEmitter, it already handles that internally, I remember using it with native animated.

As for events sent when there are no listeners I'm pretty sure there is some logic in EventEmitter to avoid doing that.

joshyhargreaves commented:
@rh389 it looks like @janicduplessis is right: the resultant calls to the RCTBridge are the same before and after the change, so not a breaking change. 

Old: https://github.com/facebook/react-native/blob/master/React/Base/RCTEventDispatcher.m#L78
New: https://github.com/facebook/react-native/blob/master/React/Modules/RCTEventEmitter.m#L49

Interesting that the integration tests seem to be failing on this assertion, any ideas about this? https://github.com/joshyhargreaves/react-native/blob/sublclass_RCTEventEmitter_in_RCTUIManager/React/Modules/RCTUIManager.m#L320

Thanks :).

joshyhargreaves commented on React/Modules/RCTUIManager.m:
```diff
@@ -287,7 +290,7 @@ - (void)invalidate
     self->_viewRegistry = nil;
     self->_bridgeTransactionListeners = nil;
-    self->_bridge = nil;
+    [self setBridge:nil];
```
This looks suspicious actually. I think this should probably just be a call to super?

joshyhargreaves commented:
Running the UIExplorerIntegrationTests locally off of master are failing for me. 

'testSimpleSnapshotTest' & 'testWebSocketTest'. The same tests seem to be failing for the travis-ci
for this feature branch.

janicduplessis commented:
Yea some tests are broken on master at the moment, sorry about that. Should be fixed Monday.

joshyhargreaves commented:
No problem, I'd just seen that was the case myself :).

janicduplessis commented:
Master is fixed now if you want to try again

rh389 commented:
Bit surprising that 'scoped' events get fired off that way through a shared dispatcher - so they're not really scoped at all? Anyway, yes you're right, no BC.

Pretty sure you'll still hit that `RCTLogWarn` on the sending side though... https://github.com/facebook/react-native/blob/master/React/Modules/RCTEventEmitter.m#L54

janicduplessis commented:
Yea I think it was just a first step towards making events scoped, it was never finished.

joshyhargreaves commented:
@rh389 I added the logic to avoid hitting the RCTLogWarn.

Interestingly I noticed a couple of other modules were actually hitting RCTLogWarn and do not have this internal 'hasObservers' logic: https://github.com/facebook/react-native/blob/master/Libraries/WebSocket/RCTWebSocketModule.m#L120, for example. 
The error was logged to the console, but no 'red screen'. 
Not sure exactly what the guidance is here, but we can probably only presume it's not to hit the RCTLogWarn from the modules!

@janicduplessis I rebased on master actually, but more tests seem to be broken!

Are there any code paths in /Examples which would the [RCTUIManager startObserving] actually to be called? I ran the integration tests, for the example, but that code path is never hit in this particular module.

rh389 commented:
I think there might be an issue there - just at a glance, if listeners are only added via the global `DeviceEventEmitter`, it looks like the `_listenerCount` on the module instance doesn't get incremented and so the event won't be fired.

joshyhargreaves commented:
Thanks, I'll do some testing.

hramos commented:
What's the latest status on this PR?

joshyhargreaves commented:
@hramos, sorry for the delay in responding!

It looks like we're now suppressing this warning this PR was fixing: https://github.com/facebook/react-native/blob/master/React/Modules/RCTUIManager.m#L239. So there perhaps is no need for this PR now unless we want to remove the deprecated API in question?

Assuming not, feel free to close, or I can close.

rh389 commented:
I wonder, does the RN team intend to finish the work towards module scoped events? If not maybe use of the global emitter should be un-deprecated? 

There are only global events on Android, and iOS has the weird behaviour I mentioned above - where global listeners don't count in the eyes of module emitters, so events may not fire.

I don't really see the point of the currently recommended iOS way, which I'm sure is just unfinished. It gives the illusion of module scoping (which would be good, I think) but in fact everything still goes via the global emitter(s) anyway.

shergin commented:
cc @javache Conceprually, this looks good to me! Are there any downsides to convert RCTUIManager to RCTEventEmitter?
I would only suggest to make `_bridge` `@protected` `ivar`, so it will make the code bit faster and the diff is smaller.

joshyhargreaves commented:
Rebasing against master.

joshyhargreaves commented:
@shergin: regarding downsides to converting `RCTUIManager` to `RCTEventEmitter`, @rh389 expressed some concern that the listener in `RCTEventEmitter` [listener count](https://github.com/facebook/react-native/blob/master/React/Modules/RCTEventEmitter.m#L48) may not be incremented (https://github.com/facebook/react-native/pull/12209#issuecomment-277706347). And therefore the event might never fire. 

I'll pick this up back and rebase and do some testing to test this.

joshyhargreaves commented:
I checked this PR a while ago and it looks like @rh389 ‘s fears were right about a lack of listeners being registered for the events in question, which means the PR wouldn’t function as intended!
